File: 6-ZigzagConversion/6-ZigzagConversion.py
# Last updated: 12/6/2025, 5:55:38 am
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        rows = [[] for _ in range(numRows)]
        n = len(s)

        row = 0
        col = 0
        reverse = False

        for i in range(n):
            rows[row].append(s[i])
            if (row+(not reverse)) % numRows == 0:
                if reverse:
                    reverse = False
                    row += 1
                else:
                    col += 1
                    reverse = True
                    row -= 1
            elif reverse:
                col += 1
                row -= 1
            else:
                row += 1
        
        ret = ''
        for row in rows:
            for col in row:
                ret += col
        
        return ret

File: 6-ZigzagConversion/6-ZigzagConversion.cpp
// Last updated: 12/6/2025, 5:55:44 am
class Solution {
public:
    string convert(string s, int numRows){
        if (s.length() <= 0) return "";
        if (s.length() == 1 || s.length() == 2 || numRows == 1) return s;
        string word = "";
        for (int i=0; i<numRows; i++){
            if (i == 0 || i == numRows-1){
                int tmp = i;
                while (true){
                    word.push_back(s[tmp]);
                    tmp+=(numRows+(numRows-2));
                    if (tmp > s.length()-1) break;
                }
            }
            else{
                int sum = 2;
                int ans = i;
                bool first = 1;
                while (true){
                    if (sum % 2 == 0){
                        if (first){
                            word.push_back(s[i]);
                            first = 0;
                        }
                        ans += ((numRows - (i+1))*2);
                    }else{
                        ans += i+1+(i-1);
                    }
                    if (ans > s.length()-1) break;
                    word.push_back(s[ans]);
                    sum++;
                }
            }
        }
        return word;
    }
};